### PR TITLE
General content update

### PR DESCRIPTION
**Pull request by joshring**
- Updated favicon to remove circle, enlarge text, C is blue, 3 is purple
- Updated landing page for the docs to try to convey a bit more of the project's goals and visually pop
- Updated arrays to add multi-dimensional fixed arrays, placed near the bottom as not key, but still helpful
- Updated optionals page to add linking to the advanced optionals page and introduce the content on each page
- Updated types page to improve the Enum, Optional and Result and some other minor things

image
image

### CODE REVIEW

File: src/content/docs/references/docs/arrays.md [hunk 1]
```diff
@@ -281,3 +282,58 @@ fn void test()
 }
 ```
 
+## Fixed multi-dimensional arrays
```
Comment (lerno) [inline src/content/docs/references/docs/arrays.md]: Fixed size, or people might think that it's saying "we fixed it!" :D

File: src/content/docs/references/docs/types.md [hunk 2]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
-    struct Foo { int x; }
+// ✅ def for each type used with a generic module.
+def IntMyList = MyList(<Foo>);
+MyListFoo working_example;
 
-    // Using def - usually recommended:
-    def IntMyList = MyList(<int>);
-    IntMyList abc;
+// ❌ An inline type definition will give an error.
+MyList<Foo> failing_example = MyList(<Foo>);
```
Comment (lerno) [inline src/content/docs/references/docs/types.md:370]: Unless in a type definition or macro

File: src/content/docs/references/docs/types.md [hunk 3]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
+    RUNNING,
+    TERMINATED
+}
 
-## Enum
+// Access enum values via:
+State current_state = State.WAITING;
+```
+The access requires referencing the `enum`'s name as `State.WAITING` because 
+an emum like `State` is a separate namespace by default, just like C++'s class `enum`.
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: typo: emum -> enum

File: src/content/docs/references/docs/types.md [hunk 4]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
+    int x;
+    int y;
+}
 
-When an enum is used in where the type can be inferred, like in case-clauses or in variable assignment, it is allowed to drop the enum name:
+enum State : int (
+    String   desc,
+    bool     active,
+    Position pos,
+)
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: Keep this declaration single line.

File: src/content/docs/references/docs/types.md [hunk 5]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
 
-### Enum associated values
+```c3
+int! i;
+i = 5; // Assigning a real value to i.
+i = IOResult.IO_ERROR?; // Assigning an optional result to i.
+```
 
-It is possible to associate each enum value with a static value.
+Only variables, function arguments and function returns may be Optionals.
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: "function arguments" is confusing here. "temporary expressions" is the most accurate, or just "expressions"

File: src/content/docs/references/docs/types.md [hunk 6]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
-    enum State : int 
-    {
-      PENDING,
-      RUNNING,
-      TERMINATED
-    }
+### Enum associated values
 
-Enum constants are namespaces by default, just like C++'s class enums. So accessing the enums above would for example use `State.PENDING` rather than `PENDING`.
+It is possible to associate each enum value one or more a static values.
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: "with one or more"

File: src/content/docs/references/docs/types.md [hunk 7]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
+```
 
-    fn void test(State s) { ... }
-
-    ...
-
-    test(RUNNING); // State.RUNNING is inferred
+### Enum type inference
 
+When an `enum` is used in where the type can be inferred, like switch case-clauses or in variable assignment, it is allowed to drop the enum name:
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: "in where" -> "where"
"it is allowed to drop the enum name" -> "the enum name is not required"

File: src/content/docs/references/docs/types.md [hunk 8]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
+test(State.RUNNING); // Uses enum constant.
+```
 
-    test(RUNNING); // Ambiguous
-    test(test::RUNNING); // Uses global.
-    test(State.RUNNING); // Uses enum constant.
+## Optional Type
 
+An [Optional type](../optionals/) is created by taking a type and appending `!`. 
+An Optional type behaves like a tagged union containing either the
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: "tagged union containing" => "tagged union, containing"

File: src/content/docs/references/docs/types.md [hunk 9]
```diff
@@ -315,125 +315,184 @@ Distinct types is a kind of type alias which creates a new type that has the sam
...
 
-    io::printfn("Currently the process is %s", s.state_desc);
-    if (s.active) do_something();
 
-## Faults
+### Optional Excuses are of type Fault
 
-`fault` defines a set of optional result values, that are similar to enums, but are used for 
-optional returns.
+When an [Optional](../optionals/) is does not contain a result, it is empty, and has an Excuse, which is of type `fault`.
```
Comment (lerno) [inline src/content/docs/references/docs/types.md]: "is does" => "does"

**Comment by joshring** [comment]
Thanks for the comments, all in now

**Comment by joshring** [comment]
Added your answer from discord:
image

File: src/content/docs/references/docs/examples.md [hunk 10]
```diff
@@ -565,7 +565,11 @@ macro long @fib(long $n)
 const long FIB19 = @fib(19); 
 // Same as const long FIB19 = 4181;
 ```
+:::note
+C3 macros are not intended for intense compile time work there are other mechanisms, such as `$exec` for that purpose.
```
Comment (lerno) [inline src/content/docs/references/docs/examples.md]: Maybe rephrase it:
> C3 macros are designed to provide a replacement for C preprocessor macros. They extend such macros by providing compile time evaluation using constant folding, which offers an IDE friendly, limited, compile time execution.
> 
> However, if you are doing more complex compile time code generation it is recommended to use `$exec` and related techniques to generate code in external scripts instead.

**Comment by joshring** [comment]
Updated as above :+1:

**Comment by lerno** [comment]
Thank you!
